akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala

```diff
@@ -240,15 +240,16 @@ private[remote] object FlushOnShutdown {
 private[remote] class FlushOnShutdown(done: Promise[Done], timeout: FiniteDuration,
                                       inboundContext: InboundContext, associations: Set[Association]) extends Actor {
 
-  var remaining = associations.flatMap(_.associationState.uniqueRemoteAddressValue)
+  var remaining = Map.empty[UniqueAddress, Int]
 
   val timeoutTask = context.system.scheduler.scheduleOnce(timeout, self, FlushOnShutdown.Timeout)(context.dispatcher)
 
   override def preStart(): Unit = {
-    // FIXME shall we also try to flush the ordinary message stream, not only control stream?
-    val msg = ActorSystemTerminating(inboundContext.localAddress)
     try {
-      associations.foreach { a ⇒ a.send(msg, OptionVal.Some(self), OptionVal.None) }
+      associations.foreach { a ⇒
+        val acksExpected = a.sendTerminationHint(self)
+        remaining += a.associationState.uniqueRemoteAddressValue().get → acksExpected
```

Comment:
`uniqueRemoteAddressValue` may be `None`, before handshake completed, so we should skip those